File: 239-sliding-window-maximum/sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& a, int k) 
    {
        int i=0;
        vector<int> ans;
        deque<int> l;          
        for(int j=0;j<a.size();j++)
        {
            while(!l.empty() and a[j]>l.back())
            l.pop_back();

            l.push_back(a[j]);

            if(j-i+1==k)
            {
                int front=l.front();
                ans.push_back(front);
                if(a[i]==front)
                l.pop_front();
                i++;
            }
        }
        return ans;
    }
};

File: 239-sliding-window-maximum/sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int[] result = new int[nums.length - k + 1];
        Deque<Integer> deque = new ArrayDeque<>();
        for(int i = 0; i < nums.length; i++) {
            while(!deque.isEmpty() && deque.peekLast() < nums[i]) {
                deque.pollLast();
            }
            deque.offerLast(nums[i]);
            if(i - k + 1 >= 0) {
                result[i - k + 1] = deque.peekFirst();
                if(nums[i - k + 1] == deque.peekFirst()) {
                    deque.pollFirst();
                }
            }
        }
        return result;
    }
}